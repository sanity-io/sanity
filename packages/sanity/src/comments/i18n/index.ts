import {defineLocaleResourceBundle} from 'sanity'

/**
 * The locale namespace for the comments plugin
 *
 * @public
 */
export const commentsLocaleNamespace = 'comments' as const

/**
 * The default locale bundle for the comments plugin, which is US English.
 *
 * @internal
 */
export const commentsUsEnglishLocaleBundle = defineLocaleResourceBundle({
  locale: 'en-US',
  namespace: commentsLocaleNamespace,
  resources: () => import('./resources'),
})

/**
 * The locale resource keys for the comments plugin.
 *
 * @alpha
 * @hidden
 */
export type {CommentsLocaleResourceKeys} from './resources'
