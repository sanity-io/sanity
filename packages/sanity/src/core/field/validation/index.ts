import {SchemaType, ObjectSchemaType} from '@sanity/types'

/** @internal */
export interface FieldValueError {
  message: string
  value: unknown
}

/** @internal */
export function getValueError(value: unknown, schemaType: SchemaType): FieldValueError | undefined {
  const {jsonType} = schemaType
  const valueType = Array.isArray(value) ? 'array' : typeof value

  if (value === null || valueType === 'undefined') {
    return undefined
  }

  if (valueType !== jsonType) {
    return {message: `Value is ${valueType}, expected ${jsonType}`, value}
  }

  if (isObjectType(schemaType) && isObjectValue(value)) {
    for (const field of schemaType.fields) {
      const fieldError = getValueError(value[field.name], field.type as SchemaType)
      if (fieldError) {
        return fieldError
      }
    }
  }

  return undefined
}

function isObjectType(schemaType: SchemaType): schemaType is ObjectSchemaType {
  return schemaType.jsonType === 'object'
}

function isObjectValue(value: unknown): value is Record<string, unknown> {
  return value !== null && !Array.isArray(value) && typeof value === 'object'
}
