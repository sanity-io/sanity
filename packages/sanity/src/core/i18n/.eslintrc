{
  "rules": {
    "@sanity/i18n/no-i18next-import": 0
  }
}