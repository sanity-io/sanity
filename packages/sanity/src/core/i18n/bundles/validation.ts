/* eslint sort-keys: "error" */
import {defineLocaleResourceBundle, defineLocalesResources} from '../helpers'
import {validationLocaleNamespace} from '../localeNamespaces'

/**
 * The string resources for validation.
 *
 * @internal
 */
const validationLocaleStrings = defineLocalesResources('validation', {
  /** Array must have exactly "$wantedLength" items, but has more/less */
  'array.exact-length': 'Must have exactly {{wantedLength}} items',
  /** Portable Text array must have exactly "$wantedLength" blocks, but has more/less */
  'array.exact-length_blocks': 'Must have exactly {{wantedLength}} blocks',
  /** Array item is a duplicate, but array wants only unique items */
  'array.item-duplicate': "Can't be a duplicate",
  /** Array has more than the maximum of "$maxLength" items */
  'array.maximum-length': 'Must have at most {{maxLength}} items',
  /** Portable Text array has more than the maximum of "$maxLength" items */
  'array.maximum-length_blocks': 'Must have at most {{maxLength}} blocks',
  /** Array has less than the minimum of "$minLength" items */
  'array.minimum-length': 'Must have at least {{minLength}} items',
  /** Portable Text array has less than the minimum of "$minLength" blocks */
  'array.minimum-length_blocks': 'Must have at least {{minLength}} blocks',
  /** Date is not valid or not in the correct format (ISO-8601) */
  'date.invalid-format': 'Must be a valid ISO-8601 formatted date string',
  /** Date is later than the given maximum date "$maxDate" */
  'date.maximum': 'Must be at or before {{maxDate}}',
  /** Date is earlier than the given minimum date "$minDate" */
  'date.minimum': 'Must be at or after {{minDate}}',
  /** A value of incorrect type is found, eg found `number` instead of `string` */
  'generic.incorrect-type': 'Expected type "{{expectedType}}", got "{{actualType}}"',
  /** Value is not one of the values specifically allowed */
  'generic.not-allowed': 'Value did not match any allowed values',
  /** Value "$givenValue" is not one of the values specifically allowed */
  'generic.not-allowed_hint': 'Value "{{hint}}" did not match any allowed values',
  /** A value is expected, but none is provided */
  'generic.required': 'Required',
  /** Number is less than the given minimum threshold value "$threshold" */
  'number.greater-than': 'Must be greater than {{threshold}}',
  /** Number is greater than the given maximum threshold value "$threshold" */
  'number.less-than': 'Must be less than {{threshold}}',
  /** Number is higher than the given maximum value "$maxNumber" */
  'number.maximum': 'Must be lower than or equal to {{maxNumber}}',
  /** Number has more precision (decimals) than the allowed "$limit" */
  'number.maximum-precision': 'Max precision is {{limit}}',
  /** Number is lower than the given minimum value "$minNumber" */
  'number.minimum': 'Must be greater than or equal to {{minNumber}}',
  /** Number is not an integer ("whole number") */
  'number.non-integer': 'Must be an integer',
  /** Object is missing a reference to an asset document in its `asset` field */
  'object.asset-required': 'Asset is required',
  /** Object is missing a reference to a file asset document in its `asset` field */
  'object.asset-required_file': 'File is required',
  /** Object is missing a reference to an image asset document in its `asset` field */
  'object.asset-required_image': 'Image is required',
  /** Object is not a reference to a document (eg `{_ref: 'documentId'}`) */
  'object.not-reference': 'Must be a reference to a document',
  /** Object references a document which is not published */
  'object.reference-not-published': 'Referenced document must be published',
  /** Accessibility label for closing the validation panel */
  'panel.close-button-aria-label': 'Close validation',
  /** Message shown when the validation panel is opened but there are no errors/warnings */
  'panel.no-errors-message': 'No validation errors',
  /** Title for the actual "Validation" panel/feature */
  'panel.title': 'Validation',
  /** Slug is an object, but is missing a `current` string property */
  'slug.missing-current': 'Slug must have a value',
  /** Slug is not an object (eg `{current: 'some-slug'}`) */
  'slug.not-object': 'Slug must be an object',
  /** Slug is already in use somewhere else, but needs to be unique */
  'slug.not-unique': 'Slug is already in use',
  /** String is not a valid email address */
  'string.email': 'Must be a valid email address',
  /** String has a different character length than the exact number "$wantedLength" */
  'string.exact-length': 'Must be exactly {{wantedLength}} characters long',
  /** String contains characters that are not in lowercase  */
  'string.lowercase': 'Must be all lowercase characters',
  /** String is longer than the limit of "$maxLength" characters */
  'string.maximum-length': 'Must be at most {{maxLength}} characters long',
  /** String is shorter than the limit of "$minLength" characters */
  'string.minimum-length': 'Must be at least {{minLength}} characters long',
  /** String does not match the given regular expression, but should */
  'string.regex-does-not-match': 'Does not match "{{name}}"-pattern',
  /** String matches the given regular expression, but should not */
  'string.regex-match': 'Should not match "{{name}}"-pattern',
  /** String contains characters that are not in uppercase */
  'string.uppercase': 'Must be all uppercase characters',
  /** String contains a protocol/scheme that is not allowed, eg (`ftp`, `mailto`â€¦) */
  'string.url.disallowed-scheme': 'Does not match allowed protocols/schemes',
  /** String contains a URL with a username or password specified before the host */
  'string.url.includes-credentials': 'Username/password not allowed',
  /** String is not a valid URL */
  'string.url.invalid': 'Not a valid URL',
  /** String is not an absolute URL (eg it is missing a protocol/host) */
  'string.url.not-absolute': 'Relative URLs are not allowed',
  /** String is not a relative URL (eg it contains a protocol/host) */
  'string.url.not-relative': 'Only relative URLs are allowed',
} as const)

/**
 * The i18n resource keys for the validation.
 *
 * @alpha
 * @hidden
 */
export type ValidationLocaleResourceKeys = keyof typeof validationLocaleStrings

/**
 * Locale resources for the validation namespace, eg US English locale resources.
 *
 * @beta
 * @hidden
 */
export const validationLocaleResources = defineLocaleResourceBundle({
  locale: 'en-US',
  namespace: validationLocaleNamespace,
  resources: validationLocaleStrings,
})
