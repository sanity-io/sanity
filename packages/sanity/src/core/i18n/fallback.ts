import {memoize} from 'lodash'
import {type Resource, createInstance} from 'i18next'
import type {LocaleSource} from './types'
import {studioLocaleNamespace} from './localeNamespaces'
import {defaultLocale, usEnglishLocale} from './locales'
import {isStaticResourceBundle} from './helpers'

const shouldEscape = typeof window === 'undefined' || typeof document === 'undefined'
const fallbackLocales: LocaleSource['locales'] = [
  {id: defaultLocale.id, title: defaultLocale.title},
]

/**
 * Returns a fallback source for internationalization that can be used in cases where you need
 * access to the "framework" of translation, but do not have access to an actual source.
 *
 * Memoized - only initializes the i18n instance once, and maintains the identity of the source.
 *
 * ⚠️ NOTE: This will only have the base studio strings defined - no custom bundles or plugins.
 *
 * @returns The fallback source
 * @internal
 */
export const getFallbackLocaleSource: () => LocaleSource = memoize(
  function getFallbackLocaleSource(): LocaleSource {
    const i18n = getFallbackI18nInstance()
    i18n.init()
    return {
      currentLocale: defaultLocale.id,
      locales: fallbackLocales,
      loadNamespaces: i18n.loadNamespaces,
      t: i18n.t,
    }
  },
)

function getFallbackI18nInstance() {
  // Find all core locale resource bundles we can load synchronously
  const staticResources: Resource = {[defaultLocale.id]: {}}
  const staticBundles = usEnglishLocale.bundles?.filter(isStaticResourceBundle) || []
  const namespaces = new Set<string>()
  for (const bundle of staticBundles) {
    staticResources[defaultLocale.id][bundle.namespace] = bundle.resources
    namespaces.add(bundle.namespace)
  }

  return createInstance({
    ns: Array.from(namespaces),
    defaultNS: studioLocaleNamespace,
    initImmediate: true,
    partialBundledLanguages: true,
    fallbackLng: defaultLocale.id,
    lng: defaultLocale.id,
    supportedLngs: [defaultLocale.id],
    debug: false,
    load: 'currentOnly',
    resources: staticResources,
    interpolation: {
      // If we're in a browser, assume this is running inside of the studio, eg a React app,
      // and that values returned will be escaped by the framework (eg React) automatically.
      escapeValue: shouldEscape,
    },
  })
}
