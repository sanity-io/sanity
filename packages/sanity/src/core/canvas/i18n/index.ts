import {type LocaleResourceBundle} from '../../i18n'

/**
 * The locale namespace for the canvas integration plugin
 *
 * @public
 */
export const canvasLocaleNamespace = 'canvas' as const

/**
 * The default locale bundle for the canvas integration plugin, which is US English.
 *
 * @internal
 */
export const canvasUsEnglishLocaleBundle: LocaleResourceBundle = {
  locale: 'en-US',
  namespace: canvasLocaleNamespace,
  resources: () => import('./resources'),
}

/**
 * The locale resource keys for the canvas integration plugin.
 *
 * @alpha
 * @hidden
 */
export type {CanvasLocaleResourceKeys} from './resources'
