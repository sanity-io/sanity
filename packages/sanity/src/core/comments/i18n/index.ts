import {type LocaleResourceBundle} from '../../i18n'

/**
 * The locale namespace for the comments plugin
 *
 * @public
 */
export const commentsLocaleNamespace = 'comments' as const

/**
 * The default locale bundle for the comments plugin, which is US English.
 *
 * @internal
 */
export const commentsUsEnglishLocaleBundle: LocaleResourceBundle = {
  locale: 'en-US',
  namespace: commentsLocaleNamespace,
  resources: () => import('./resources'),
}

/**
 * The locale resource keys for the comments plugin.
 *
 * @alpha
 * @hidden
 */
export type {CommentsLocaleResourceKeys} from './resources'
