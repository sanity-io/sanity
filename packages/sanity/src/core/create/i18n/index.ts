import {type LocaleResourceBundle} from '../../i18n'

/**
 * The locale namespace for the Create integration plugin
 *
 * @public
 */
export const createLocaleNamespace = 'create' as const

/**
 * The default locale bundle for the Create integration plugin, which is US English.
 *
 * @internal
 */
export const createUsEnglishLocaleBundle: LocaleResourceBundle = {
  locale: 'en-US',
  namespace: createLocaleNamespace,
  resources: () => import('./resources'),
}

/**
 * The locale resource keys for the Create integration plugin.
 *
 * @alpha
 * @hidden
 */
export type {CreateLocaleResourceKeys} from './resources'
