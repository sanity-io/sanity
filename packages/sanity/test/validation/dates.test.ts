import {describe, expect, test} from 'vitest'

import {getFallbackLocaleSource} from '../../src/core/i18n/fallback'
import {Rule} from '../../src/core/validation'

describe('date', () => {
  describe('with default format', () => {
    const context: any = {client: {}, i18n: getFallbackLocaleSource(), type: {name: 'date'}}

    test('min length constraint', async () => {
      const rule = Rule.dateTime().min('2024-01-01')
      await expect(rule.validate('2023-12-31', context)).resolves.toMatchSnapshot(
        'Must be at or after',
      )
      await expect(rule.validate('2024-01-02', context)).resolves.toHaveLength(0)
      await expect(rule.validate('2024-01-01', context)).resolves.toHaveLength(0)
    })

    test('max length constraint', async () => {
      const rule = Rule.dateTime().max('2024-01-01')
      await expect(rule.validate('2024-01-02', context)).resolves.toMatchSnapshot(
        'Must be at or before',
      )
      await expect(rule.validate('2023-12-31', context)).resolves.toHaveLength(0)
      await expect(rule.validate('2024-01-01', context)).resolves.toHaveLength(0)
    })
  })

  describe('with custom format', () => {
    const context: any = {
      client: {},
      i18n: getFallbackLocaleSource(),
      type: {name: 'date', options: {dateFormat: 'MM-DD-YYYY'}},
    }

    test('min length constraint', async () => {
      const rule = Rule.dateTime().min('2024-01-01')
      await expect(rule.validate('2023-12-31', context)).resolves.toMatchSnapshot(
        'Must be at or after',
      )
      await expect(rule.validate('2024-01-02', context)).resolves.toHaveLength(0)
      await expect(rule.validate('2024-01-01', context)).resolves.toHaveLength(0)
    })

    test('max length constraint', async () => {
      const rule = Rule.dateTime().max('2024-01-01')
      await expect(rule.validate('2024-01-02', context)).resolves.toMatchSnapshot(
        'Must be at or before',
      )
      await expect(rule.validate('2023-12-31', context)).resolves.toHaveLength(0)
      await expect(rule.validate('2024-01-01', context)).resolves.toHaveLength(0)
    })
  })
})

describe('datetime', () => {
  describe('with default format', () => {
    const context: any = {client: {}, i18n: getFallbackLocaleSource(), type: {name: 'datetime'}}

    test('min length constraint', async () => {
      const rule = Rule.dateTime().min('2024-01-01T17:31:00.000Z')
      await expect(rule.validate('2023-12-31T17:31:00.000Z', context)).resolves.toMatchSnapshot(
        'Must be at or after',
      )
      await expect(rule.validate('2024-01-02T17:31:00.000Z', context)).resolves.toHaveLength(0)
      await expect(rule.validate('2024-01-01T17:31:00.000Z', context)).resolves.toHaveLength(0)
    })

    test('max length constraint', async () => {
      const rule = Rule.dateTime().max('2024-01-01T17:31:00.000Z')
      await expect(rule.validate('2024-01-02T17:31:00.000Z', context)).resolves.toMatchSnapshot(
        'Must be at or before',
      )
      await expect(rule.validate('2023-12-23T17:31:00.000Z', context)).resolves.toHaveLength(0)
      await expect(rule.validate('2024-01-01T17:31:00.000Z', context)).resolves.toHaveLength(0)
    })
  })

  describe('with custom format', () => {
    const context: any = {
      client: {},
      i18n: getFallbackLocaleSource(),
      type: {name: 'datetime', options: {dateFormat: 'Do. MMMM YYYY'}},
    }

    test('min length constraint', async () => {
      const rule = Rule.dateTime().min('2024-01-01T17:31:00.000Z')
      await expect(rule.validate('2023-12-31T17:31:00.000Z', context)).resolves.toMatchSnapshot(
        'Must be at or after',
      )
      await expect(rule.validate('2024-01-02T17:31:00.000Z', context)).resolves.toHaveLength(0)
      await expect(rule.validate('2024-01-01T17:31:00.000Z', context)).resolves.toHaveLength(0)
    })

    test('max length constraint', async () => {
      const rule = Rule.dateTime().max('2024-01-01T17:31:00.000Z')
      await expect(rule.validate('2024-01-02T17:31:00.000Z', context)).resolves.toMatchSnapshot(
        'Must be at or before',
      )
      await expect(rule.validate('2023-12-31T17:31:00.000Z', context)).resolves.toHaveLength(0)
      await expect(rule.validate('2024-01-01T17:31:00.000Z', context)).resolves.toHaveLength(0)
    })
  })
})
