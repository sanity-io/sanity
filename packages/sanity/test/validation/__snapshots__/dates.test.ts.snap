// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`date > with custom format > max length constraint > Must be at or before 1`] = `
[
  {
    "item": {
      "message": "Must be at or before 01-01-2024",
    },
    "level": "error",
    "message": "Must be at or before 01-01-2024",
    "path": [],
  },
]
`;

exports[`date > with custom format > min length constraint > Must be at or after 1`] = `
[
  {
    "item": {
      "message": "Must be at or after 01-01-2024",
    },
    "level": "error",
    "message": "Must be at or after 01-01-2024",
    "path": [],
  },
]
`;

exports[`date > with default format > max length constraint > Must be at or before 1`] = `
[
  {
    "item": {
      "message": "Must be at or before 2024-01-01",
    },
    "level": "error",
    "message": "Must be at or before 2024-01-01",
    "path": [],
  },
]
`;

exports[`date > with default format > min length constraint > Must be at or after 1`] = `
[
  {
    "item": {
      "message": "Must be at or after 2024-01-01",
    },
    "level": "error",
    "message": "Must be at or after 2024-01-01",
    "path": [],
  },
]
`;

exports[`datetime > with custom format > max length constraint > Must be at or before 1`] = `
[
  {
    "item": {
      "message": "Must be at or before 1st. January 2024 09:31",
    },
    "level": "error",
    "message": "Must be at or before 1st. January 2024 09:31",
    "path": [],
  },
]
`;

exports[`datetime > with custom format > min length constraint > Must be at or after 1`] = `
[
  {
    "item": {
      "message": "Must be at or after 1st. January 2024 09:31",
    },
    "level": "error",
    "message": "Must be at or after 1st. January 2024 09:31",
    "path": [],
  },
]
`;

exports[`datetime > with default format > max length constraint > Must be at or before 1`] = `
[
  {
    "item": {
      "message": "Must be at or before 2024-01-01 09:31",
    },
    "level": "error",
    "message": "Must be at or before 2024-01-01 09:31",
    "path": [],
  },
]
`;

exports[`datetime > with default format > min length constraint > Must be at or after 1`] = `
[
  {
    "item": {
      "message": "Must be at or after 2024-01-01 09:31",
    },
    "level": "error",
    "message": "Must be at or after 2024-01-01 09:31",
    "path": [],
  },
]
`;
